### PR TITLE
adapt old config to new standards

### PR DESCRIPTION
**Pull request by lyninx**
depends on: https://github.com/Shopify/js-buy-sdk/pull/387

this change allows existing buy buttons, which historically used numerical shopify product IDs, to be transformed into our GraphQL API's product IDs so that they don't fall over once we roll out the updated `buy-button-js` library to existing buy button embeds.

### CODE REVIEW

File: src/components/product.js [hunk 1]
```diff
@@ -52,6 +52,22 @@ function whitelistedProperties(selectorStyles) {
   }, {});
 }
 
+function adaptConfig(config) {
```
Comment (minasmart) [inline src/components/product.js]: this isn't doing anything with a config. It's normalizing an ID. Consider a different name.

File: src/components/product.js [hunk 2]
```diff
@@ -52,6 +52,22 @@ function whitelistedProperties(selectorStyles) {
...
+  if (config.id && !isBase64(config.id)) {
+    config.id = btoa(`gid://shopify/Product/${config.id}`);
+  }
+
+  return config;
+}
+
+function isBase64(str) {
+  try {
+    return btoa(atob(str)) === str;
```
Comment (minasmart) [inline src/components/product.js]: I'm not sure this is the best way to do this.

```javascript
str = '1234'

btoa(atob(str)) === str
```

The above resolves to true.

It may make sense to have an `id` field, which is deprecated, and a `storefrontId` field, which is the new ID format.

The base64 encoding format doesn't make this a deterministic check if the string is a multiple of 4 characters long.

File: src/components/product.js [hunk 3]
```diff
@@ -461,7 +477,7 @@ export default class Product extends Component {
    * @return {String}
    */
   get onlineStoreURL() {
-    const identifier = this.handle ? this.handle : this.id;
+    const identifier = this.handle ? this.handle : atob(this.id).split('/').pop();
```
Comment (minasmart) [inline src/components/product.js]: This is probably an  anti-pattern. We should likely add the online store URL as a field on product. The domain is not always going to be the myshopify domain (which is what's configured here).

File: src/components/product.js (hunk 2 above)
Comment (lyninx) [inline src/components/product.js]: i like this `storefrontId` idea. definitely way nicer than trying to detect what type of ID was provided.

Comment (minasmart) [inline src/components/product.js]: also, the only deterministic way 😄

File: src/components/product.js (hunk 3 above)
Comment (swalkinshaw) [inline src/components/product.js]: @minasmart is this valid though? If you have a custom storefront then you might not have online store enabled.

Comment (minasmart) [inline src/components/product.js]: you might not, but there are valid use cases to go from the buy button view to the online store view. That's a behaviour that exists today that we can't deprecate.

Comment (minasmart) [inline src/components/product.js]: also, we could actually use the same logic we use for blogs to show and hide the value in the API, as opposed to assuming it's always there generating it on the client.

**Comment by lyninx** [comment]
i'll address the onlineStoreURL issue in a separate PR, once we add that to the api #402

File: src/components/product.js [hunk 4]
```diff
@@ -65,7 +75,7 @@ export default class Product extends Component {
    * @param {Object} props - data and utilities passed down from UI instance.
    */
   constructor(config, props) {
-    super(config, props);
+    super(normalizeProductId(config), props);
```
Comment (minasmart) [inline src/components/product.js]: I'd use the same pattern we used in `js-buy-sdk`. It's kind of ugly to have a method that mutates its input. Can you receive the config then do `config.id = normalizeProductId(config)`?

File: src/components/product.js [hunk 5]
```diff
@@ -163,7 +173,8 @@ export default class Product extends Component {
    * @return {Object} viewData object.
    */
   get viewData() {
-    return merge(this.model, this.options.viewData, {
+    const model = Object.assign({}, this.model);
+    return merge(model, this.options.viewData, {
```
Comment (minasmart) [inline src/components/product.js]: is `merge` different than `Object.assign`?

File: src/components/product.js (hunk 3 above)
Comment (minasmart) [inline src/components/product.js]: Let's add an issue to the storefront API feature board @swalkinshaw created about getting this added to the API so it's not lost.

File: test/unit/product.js [hunk 6]
```diff
@@ -356,7 +356,7 @@ describe('Product class', () => {
 
       it('calls fetchProduct with product id', () => {
```
Comment (minasmart) [inline test/unit/product.js]: Are you testing with a storefront id, and product id?

File: src/components/product.js (hunk 4 above)
Comment (minasmart) [inline src/components/product.js]: Also, we should probably be preferentially using `storefrontId` everywhere, and deprecate the use of `id`. This can be a different PR though.

File: src/components/product.js (hunk 5 above)
Comment (minasmart) [inline src/components/product.js]: my question is why you don't just do `Object.assign({}, this.model, this.options.viewData, { ... })`

Comment (lyninx) [inline src/components/product.js]: oh.... duh. yeah merge isnt requires any more. good call!

File: test/unit/product.js (hunk 6 above)
Comment (lyninx) [inline test/unit/product.js]: I'll stick with storefrontId for this test, and make a separate test for the normalization.

File: src/components/product.js [hunk 7]
```diff
@@ -65,6 +75,7 @@ export default class Product extends Component {
    * @param {Object} props - data and utilities passed down from UI instance.
    */
   constructor(config, props) {
+    config.id = normalizeProductId(config).id;
```
Comment (minasmart) [inline src/components/product.js]: This assignment is a noop. You're mutating config inside the function. You could do this without the assignment and have the same side effect.

Try to have the method receive data and return data without mutating anything so like:

```javascript
function normalizeProductId(config) {
   if (config.storefrontId) {
     return config.storefrontId;
   } else if (config.id) {
     return = btoa(`gid://shopify/Product/${config.id}`);
   } else {
    throw new Error('No ID was provided');
  }
 }

config.storefrontId = normalizeProductId(config);
```

Mutating inside the function is obfuscating the behaviour you're trying to represent. Also, since the storefront ID is the canonical ID that you're using, storefront ID should be what you're assigning. The ID returned by this is not a database id.

Comment (lyninx) [inline src/components/product.js]: wow, yeah not mutating the object in the function seems much cleaner. 💯 
as for `storefrontId` instead of `id`, i'm planning on addressing that by converting the id field across the whole library in a separate pr.

Comment (minasmart) [inline src/components/product.js]: 👍

**Comment by lyninx** [comment]
i'm leaving out the error that should be thrown if no ID is provided for now, since there may be early buy-buttons out in the wild which use handles, rather than IDs.

File: src/components/product-set.js [hunk 8]
```diff
@@ -127,31 +127,20 @@ export default class ProductSet extends Component {
...
-        }
-        return Promise.resolve([]);
+      method = this.props.client.fetchCollectionByHandle(this.handle).then((collection) => {
+        this.id = collection.id;
+        return this.props.client.fetchCollectionWithProducts(this.id);
       });
     }
-    return method;
+    return method.then((collection) => {
+      return Promise.resolve(collection.products);
```
Comment (lyninx) [inline src/components/product-set.js]: that it is. will fix.

File: src/components/product.js [hunk 9]
```diff
@@ -65,6 +75,7 @@ export default class Product extends Component {
    * @param {Object} props - data and utilities passed down from UI instance.
    */
   constructor(config, props) {
+    config.id = normalizeProductId(config);
```
Comment (lyninx) [inline src/components/product.js]: i'm planning on handling that whole id cleanup as a separate PR

File: src/components/product.js [hunk 10]
```diff
@@ -461,8 +472,7 @@ export default class Product extends Component {
    * @return {String}
    */
   get onlineStoreURL() {
-    const identifier = this.handle ? this.handle : this.id;
-    return `https://${this.props.client.config.domain}/products/${identifier}${this.onlineStoreQueryString}`;
+    return `${this.model.onlineStoreUrl}${this.onlineStoreQueryString}`;
```
Comment (lyninx) [inline src/components/product.js]: it adds query params to the url such as the referring channel / domain, as well as which specific variant is chosen

Comment (minasmart) [inline src/components/product.js]: ah, cool.
